src/rebar_erlc_compiler.erl

```diff
@@ -147,164 +146,84 @@ doterl_compile(Config, Dir, OutDir, MoreSources, ErlOpts) ->
-u_add_element(Elem, [E1|Set])     -> [E1|u_add_element(Elem, Set)];
-u_add_element(Elem, [])           -> [Elem].
-
--spec include_path(file:filename(),
-                   rebar_state:t()) -> [file:filename(), ...].
-include_path(Source, Config) ->
-    ErlOpts = rebar_state:get(Config, erl_opts, []),
-    Dir = filename:join(rebar_utils:droplast(filename:split(filename:dirname(Source)))),
-    lists:usort([filename:join(Dir, "include"), filename:dirname(Source)]
-                ++ proplists:get_all_values(i, ErlOpts)).
-
--spec needs_compile(file:filename(), file:filename(),
-                    list(), [string()]) -> boolean().
-needs_compile(Source, Target, Opts, Parents) ->
-    TargetLastMod = filelib:last_modified(Target),
-    F = fun(I) -> source_changed(TargetLastMod, I) end,
-    lists:any(F, [Source] ++ Parents) orelse opts_changed(Opts, Target).
-
-source_changed(TargetLastMod, I) -> TargetLastMod < filelib:last_modified(I).
-
-opts_changed(NewOpts, Target) ->
-    case compile_info(Target) of
-        {ok, Opts} -> lists:sort(Opts) =/= lists:sort(NewOpts);
-        _          -> true
-    end.
-
-compile_info(Target) ->
-    case beam_lib:chunks(Target, [compile_info]) of
-        {ok, {_mod, Chunks}} ->
-            CompileInfo = proplists:get_value(compile_info, Chunks, []),
-            {ok, proplists:get_value(options, CompileInfo, [])};
-        {error, beam_lib, Reason} ->
-            ?WARN("Couldn't read debug info from ~p for reason: ~p", [Target, Reason]),
-            {error, Reason}
+erl_first_files(Config, NeededErlFiles) ->
+    %% NOTE: rebar_config:get_local perhaps?
+    ErlFirstFilesConf = rebar_state:get(Config, erl_first_files, []),
+    %% NOTE: order of files in ErlFirstFiles is important!
+    [File || File <- ErlFirstFilesConf, lists:member(File, NeededErlFiles)].
+
+%% Get subset of SourceFiles which need to be recompiled, respecting
+%% dependencies induced by given graph G.
+needed_files(G, ErlOpts, Dir, OutDir, SourceFiles) ->
+    lists:filter(fun(Source) ->
+                         Target = target_base(OutDir, Source) ++ ".beam",
+                         Opts = [{outdir, filename:dirname(Target)}
+                                ,{i, filename:join(Dir, "include")}] ++ ErlOpts,
+                         digraph:vertex(G, Source) > {Source, filelib:last_modified(Target)}
+                              orelse opts_changed(Opts, Target)
+                 end, SourceFiles).
+
+opts_changed(Opts, Target) ->
+    Basename = filename:basename(Target, ".beam"),
+    Dirname = filename:dirname(Target),
+    ObjectFile = filename:join([Dirname, Basename]),
+    case code:load_abs(ObjectFile) of
+        {module, Mod} ->
+            Compile = Mod:module_info(compile),
+            lists:sort(Opts) =/= lists:sort(proplists:get_value(options,
+                                                                Compile,
+                                                                undefined));
```

Comment:
See my comments to this particular source. I don't know why they are not displayed here.